src/main/scala/org/broadinstitute/dsde/agora/server/webservice/AddHandler.scala

```diff
@@ -53,4 +53,69 @@ class AddHandler(authorizationProvider: AuthorizationProvider) extends Actor {
     val method = agoraBusiness.insert(agoraEntity.copy(createDate = Option(new DateTime())), username)
     context.parent ! RequestComplete(spray.http.StatusCodes.Created.intValue, method)
   }
+
+}
+
+case class MethodImportResolver(username: String, business: AgoraBusiness, authorizationProvider: AuthorizationProvider) {
+
+  def importResolver(importUri: String): WdlSource = {
+    ImportResolverHelper.validateUri(importUri)
+    val method = ImportResolverHelper.resolve(importUri, business, username)
+    val authorizedMethod = authorizationProvider.filterByReadPermissions(method, username)
+    authorizedMethod match {
+      case x: Some[AgoraEntity] => x.get.payload.get
+      case None => throw new SyntaxError(s"Can't resolve import: $importUri")
+    }
+  }
+}
+
+object ImportResolverHelper {
+
+  private val methodScheme = "methods://"
+
+  def resolve(uri: String, business: AgoraBusiness, username: String): Option[AgoraEntity] = {
+    val withoutScheme = removeScheme(uri)
+    val parts = uriParts(withoutScheme)
+    val namespace = parts(0)
+    val name = parts(1)
+    val id = parts(2).toInt
+    val method = business.findSingle(namespace, name, id, Seq(AgoraEntityType.Task, AgoraEntityType.Workflow), username)
+    method
+  }
+
+  def validateUri(importUri: String) = {
+    if (!importUri.startsWith(methodScheme)) {
+      throw new SyntaxError(s"Unsupported import uri: $importUri. Must start with $methodScheme.")
+    }
+    val withoutScheme = removeScheme(importUri)
+    if (!hasThreeParts(withoutScheme)) {
+      throw new SyntaxError(s"Malformed methods uri. Must contain three parts, separated by '.': $importUri")
+    }
+    val parts = importUri.split("\\.")
+    if (!isIdInt(parts(2))) {
+      throw new SyntaxError(s"Malformed methods uri. SnapshotId must be integer: $importUri")
+    }
+  }
+
+  private def removeScheme(uri: String): String = {
+    uri.substring(methodScheme.length)
+  }
+
+  private def hasThreeParts(uriWithoutScheme: String) = {
+    val parts = uriParts(uriWithoutScheme)
+    parts.size == 3
+  }
+
+  private def isIdInt(idStr: String): Boolean = {
+    try {
+      idStr.toInt
+      true
+    } catch {
+      case e: NumberFormatException => false
+    }
```

Comment:
why not idStr.forall(_.isDigit)? Might be better than parsing the string to an int that we are just throwing away